### ISSUE TITLE
Failing to Save Checkpoints with saves_per_epoch

### ISSUE THREAD

**Issue by jghawaly** [axolotl-ai-cloud/axolotl#1618]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

I expected it to save checkpoints 8 times per epoch for a total of 32 checkpoint saves.

### Current behaviour

It only saved the last 4 checkpoints at step 784, 812, 840, and 868. There are a total of 868 steps in this training run (217 per epoch).

### Steps to reproduce

Using the provided config and running the following command: 

> accelerate launch -m axolotl.cli.train finetune-llama38b.yaml

### Config yaml

```yaml
base_model: meta-llama/Meta-Llama-3-8B
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer

load_in_8bit: false
load_in_4bit: true
strict: false
eval_sample_packing: false
datasets:
  - path: json
    type: alpaca
    data_files: /home/admin/CODE/data/datasets/train.jsonl
dataset_prepared_path: /home/admin/CODE/data/datasets/prepared_datasets
val_set_size: 0.2
output_dir: ./qlora-out

sequence_len: 8000
sample_packing: true
pad_to_sequence_len: true

adapter: qlora
lora_model_dir:
lora_r: 64
lora_alpha: 64
lora_dropout: 0.05
lora_target_linear: true

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 8
micro_batch_size: 1
num_epochs: 4
optimizer: paged_adamw_32bit
lr_scheduler: cosine
learning_rate: 0.0002

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true
s2_attention:

warmup_steps: 100
evals_per_epoch: 8
saves_per_epoch: 8
debug:
deepspeed:
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
   pad_token: <|end_of_text|>
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12.3

### axolotl branch-commit

main/2147cf6837e2b90a2ea7045262083cbb0da03858

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by winglian** [comment]
there is an option `save_total_limit` that defaults to `4`. This is set so your hard drive doesn't just fall over with 10s-100s of checkpoints. You can change this to some same value that's right for oyu.
